qemu/tests/msi_change_flag.py

```diff
@@ -0,0 +1,123 @@
+import logging
+import os
+import re
+import ctypes
+
+from avocado.utils import crypto, process
+from virttest import utils_misc
+from virttest import utils_test
+from provider import win_dev
+from virttest import error_context
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    vhost is no longer disabled when guest does not use MSI-X.
+    The vhostforce flag is no longer required.
+
+    1) Start guest with different NIC option
+    2) Check virtio device's irp number,irq number should be greater than one.
+    3) Disable msi of guest
+    4) Reboot guest, check if msi is disabled and irq number should be equal to 1.
+    5) Check network and vhost process (transfer data).
+    6) Check md5 value of both sides.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    tmp_dir = params["tmp_dir"]
+    filesize = int(params.get("filesize"))
+    dd_cmd = params["dd_cmd"]
+    delete_cmd = params["delete_cmd"]
+    file_md5_check_timeout = int(params.get("file_md5_check_timeout"))
+
+    def get_file_md5sum(file_name, session, timeout):
+        """
+        return: Return the md5sum value of the guest.
+        """
+        logging.info("Get md5sum of the file:'%s'", file_name)
+        s, o = session.cmd_status_output("md5sum %s" % file_name, timeout=timeout)
+        if s != 0:
+            test.error("Get file md5sum failed as %s" % o)
+        return re.findall(r"\w{32}", o)[0]
+
+    login_timeout = int(params.get("login_timeout", 360))
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_serial_login()
+
+    if params.get("os_type") == "linux":
+        error_context.context("Check the pci msi in guest", logging.info)
+        pci_id = session.cmd("lspci |grep Eth |awk {'print $1'}").strip()
+        status = session.cmd("lspci -vvv -s %s|grep MSI-X" % pci_id).strip()
+        enable_status = re.search(r'Enable\+', status, re.M | re.I)
+        if enable_status.group() == "Enable+":
+            error_context.context("Disable pci msi in guest", logging.info)
+            utils_test.update_boot_option(vm, args_added="pci=nomsi")
+            session_msi = vm.wait_for_serial_login(timeout=login_timeout)
+            pci_id = session_msi.cmd("lspci |grep Eth |awk {'print $1'}").strip()
+            status = session_msi.cmd("lspci -vvv -s %s|grep MSI-X" % pci_id).strip()
+            session_msi.close()
+            change_status = re.search(r'Enable\-', status, re.M | re.I)
+            if change_status.group() != "Enable-":
+                test.fail("virtio device's statuts is not correct")
+        elif enable_status.group() != "Enable+":
+            test.fail("virtio device's statuts is not correct")
+    else:
+        def irq_check(session, device_name, devcon_folder):
```

Comment:
We always define all functions at the beginning under `def run`, please not inline this.